components/cdc/src/channel.rs

```diff
@@ -125,13 +201,63 @@ impl Drain {
                 });
                 let resps = batcher.build();
                 let last_idx = resps.len() - 1;
+                // Events are about to be sent, free pending events memory counter.
+                memory_quota.free(bytes as _);
                 stream::iter(resps.into_iter().enumerate().map(move |(i, e)| {
                     // Buffer messages and flush them at once.
                     let write_flags = WriteFlags::default().buffer_hint(i != last_idx);
                     GrpcResult::Ok((e, write_flags))
                 }))
             })
-            .flatten()
+            .flatten();
+        DrainGrpcMessage { stream, closed }
+    }
+}
+
+struct DrainGrpcMessage<S: Stream + Unpin> {
+    stream: S,
+    closed: Arc<AtomicBool>,
+}
+
+impl<S: Stream + Unpin> Stream for DrainGrpcMessage<S> {
+    type Item = S::Item;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::new(&mut self.stream).poll_next(cx)
+    }
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.stream.size_hint()
+    }
+}
+
+impl<S: Stream + Unpin> Drop for DrainGrpcMessage<S> {
+    // Must poll all pending messages to free memory quota.
+    fn drop(&mut self) {
+        // Receiver maybe dropped before senders, so set closed flag to prevent
+        // futher messages.
+        self.closed.store(true, Ordering::Release);
+        std::sync::atomic::compiler_fence(Ordering::SeqCst);
+        let start = Instant::now();
+        let mut drain = Box::pin(async {
+            loop {
+                let mut next = self.stream.next();
+                let item = futures::future::poll_fn(|cx| {
+                    match Pin::new(&mut next).poll(cx) {
+                        // In case it is dropped before senders.
+                        Poll::Pending => Poll::Ready(None),
+                        t => t,
+                    }
+                })
+                .await;
+                if item.is_none() {
+                    break;
```

Comment:
If `next` yields a `Poll::Pending`, will the loop break here? If so, why is it acceptable?